### DISCUSSION TITLE
LocatorJS Support - SWC Plugin

### DISCUSSION THREAD

**Discussion by stobias123** [vercel/next.js#64591] [Ideas]
### Goals

1. Support [locatorjs](https://www.locatorjs.com/) with the latest version of next.
2. Create a SWC plugin 

### Background

I recently came across the locatorjs project, and I really dig it. Quick and easy way to go directly to the component you're trying to edit. Unfortunately, I'm pretty new to next, and as such on the latest version, and this plugin doesn't work with the latest version of next (see linked discussion below).

I've spent the last week or so (fruitlessly) trying to create an SWC plugin so that it can be used with the newest version of next. The author of the project has also tried [here](https://github.com/infi-pc/locatorjs/pull/115/files) and has an open issue [here](https://github.com/infi-pc/locatorjs/issues/102) but has also been unsuccessful. I think the linked plugin is actually pretty close to all that is needed (adding some metadata to files) - but might need someone more experienced to push it over the finish line.

### Proposal

It looks like the next team has some experience migrating babel to swc components, and I really think it would be awesome to add locatorjs to the list of [supported components](https://nextjs.org/docs/architecture/nextjs-compiler#supported-features).

I'd be happy to share the code I have, it's fairly close to the original authors stuff linked above.

**Comment by trulyronak** [comment]
Would heavily +1 this proposal! Next.js + Locator would be amazing - I really don't like that I have to give up `next/font` just to use locatorjs, and locatorjs makes _any_ react/next style project better to work with.

**Comment by royk-legit** [comment]
LocatorJS worked for me on NextJS  for months, until the last version of React Developer tools extension came (version 5.3.1). 
Since then it doesn't work :/

**Comment by pitops** [comment]
LocatorJS is super helpful. Would really love for this to keep working.

**Comment by curlyz** [comment]
:) sadly going back to next 13 work for me, worth it

**Comment by MatteoGauthier** [comment]
What's the status of this discussion ?

**Comment by xxk8** [comment]
I really want to know how it is currently. If I follow the method of LocatorJS, **swc** will be gone, which makes me feel very uncomfortable

**Reply by curlyz** [reply]
I will give you 3 magic words that work with latest version of NextJS. Are you ready :)

[**code-inspection-plugin**](https://www.npmjs.com/package/code-inspector-plugin)

Found this gem from dify/web codebase.

**Reply by xxk8** [reply]
Thank you for your reply. This is not available on Next15. I have asked the group owner.
![WechatIMG949](https://github.com/user-attachments/assets/c2562545-03fa-4d87-b204-db9e25484681)

**Reply by curlyz** [reply]
I'm literally using it right now, as NextJS just release new version today

next.config.js
```
// const { codeInspectorPlugin } = require('code-inspector-plugin')

import { codeInspectorPlugin } from 'code-inspector-plugin'
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'standalone',
  webpack: (config, { dev, isServer }) => {
    config.plugins.push(codeInspectorPlugin({ bundler: 'webpack' }))
    if (dev && !isServer) {
    }

    console.log('next.config.js/webpack.plugins', config.plugins)
    return config
  },
}
export default nextConfig
// module.exports = nextConfig

```

**Reply by xxk8** [reply]
OK, thank you very much! I will try it again tomorrow morning!

**Reply by anishFE** [reply]
It is working , thank you so much !! @curlyz

**Reply by tomups** [reply]
Unfortunately this doesn't work with turbopack

**Reply by silvioprog** [reply]
I've been trying to use `code-inspection-plugin` in a huge project but getting the following error, and since the project is huge, I have no idea how to fix it:

![Screenshot From 2025-03-04 22-26-26](https://github.com/user-attachments/assets/45d42448-a45e-419d-a9f6-903964da9b73)

**Reply by curlyz** [reply]
It look like there is problem with your code, not related to the plugin.

**Reply by silvioprog** [reply]
> It look like there is problem with your code, not related to the plugin.

But the (legacy) code works fine when I remove the plugin

**Reply by sian022** [reply]
@curlyz, Thanks! It works and displays the components properly! However, when I click, it doesn't open in my IDE. I wonder why that is?

Network logs when trying to click to open in IDE:
![image](https://github.com/user-attachments/assets/dc9ebf6a-b9dc-4c61-8c4c-8e87b45ec1a1)

**Comment by nabinkatwal7** [comment]
any updates in this discussion?

**Reply by LAWTED** [reply]
You may try https://github.com/zh-lx/code-inspector, it works for me.

**Reply by nabinkatwal7** [reply]
a separate package for code inspection isn't gonna float with my dev team. I was just looking if locator js supports SWC anytime soon

**Reply by C-stle** [reply]
I successfully applied Code Inspector on Next.js 15.5. Try configuring it like this.
<img width="458" height="344" alt="스크린샷 2025-09-05 15 31 32" src="https://github.com/user-attachments/assets/08eebfec-975c-4bda-8090-919ddbea3ed7" />

**Reply by nabinkatwal7** [reply]
thanks!

**Comment by thedv91** [comment]
Late to this, but I did end up writing the SWC plugin this thread was after — narrower than what LocatorJS needs, so worth being upfront about it.

It stamps one attribute per JSX element and stops there: `<main data-tsd-source="src/app/page.tsx:7:5">`. No overlay, no runtime, no server to open your editor — just the position sitting in the markup for whatever wants to read it. If you want click-to-IDE working out of the box, code-inspector further up this thread already does that and I'd point you there. This is for when you already have your own devtools and only need the data.

https://github.com/thedv91/swc-plugin-jsx-source-attrs

```ts
// next.config.ts
experimental: { swcPlugins: [["swc-plugin-jsx-source-attrs", {}]] }
```

Turbopack and webpack both.

**If you run the React Compiler, use the loader instead of the plugin.** This is the part I'd get wrong if I were reading the thread rather than writing this: the compiler rebuilds the JSX tree with no source spans for anything it recognises as a component, so every position is gone before an SWC plugin — which runs *after* the host's transforms — ever gets asked. On one app that was 346 of 389 files reporting the bare path with no line or column. Nothing downstream can recover it; with the compiler on, even `jsxDEV` receives `void 0` where its `source` argument goes.

The loader runs ahead of the pipeline, splicing the attribute into the source itself, where it survives as an ordinary string literal:

```ts
// next.config.ts
turbopack: {
  rules: { "*.tsx": { loaders: [{ loader: "swc-plugin-jsx-source-attrs/loader", options: {} }] } },
},
```

Same attribute, same options, and it takes a plain webpack rule too. Measured on that app: 28 of 28 elements on a route carry a full position, against 2 of 28 through the plugin — and cold compile went from 6.4–6.8s to 4.4–4.8s, so it isn't paying for the privilege either.

For the click-to-IDE half, this pairs with [TanStack Devtools](https://tanstack.com/devtools). You install and mount `@tanstack/react-devtools`, the attribute above is what its source inspector reads, and the rest is config — no route handler, nothing extra compiled into production builds:

```ts
// next.config.ts
async redirects() {
  if (process.env.NODE_ENV !== "development") return [];
  return [{
    source: "/__tsd/open-source",
    permanent: false,
    has: [{ type: "query", key: "source", value: "(?<file>.+):(?<line>\\d+):(?<column>\\d+)" }],
    destination: "/__nextjs_launch-editor?file=:file&line1=:line&column1=:column",
  }];
},
```

A redirect rather than a rewrite — `__nextjs_launch-editor` is served by dev middleware that runs ahead of the router, so a rewritten URL never reaches it and 404s. Checked on Next 16.3.1 with Turbopack: hold Shift+Alt+Ctrl, hover, click, and the editor opens at the right line.

Full setup, including the client component and the failure modes that look like a dead hotkey: https://github.com/thedv91/swc-plugin-jsx-source-attrs#tanstack-devtools
